types/middleware.d.ts

```diff
@@ -0,0 +1,30 @@
+import * as http from 'http'
+import * as http2 from 'http2'
+import * as https from 'https'
+
+import { FastifyInstance } from '../fastify'
+import { FastifyRequest } from './request'
+import { FastifyReply } from './reply'
+
+export type FastifyMiddleware<
+  RawServer extends (http.Server | https.Server | http2.Http2Server | http2.Http2SecureServer) = http.Server, 
+  RawRequest extends (http.IncomingMessage | http2.Http2ServerRequest) = RawServer extends http.Server | https.Server ? http.IncomingMessage : http2.Http2ServerRequest, 
+  RawReply extends (http.ServerResponse | http2.Http2ServerResponse) = RawServer extends http.Server | https.Server ? http.ServerResponse : http2.Http2ServerResponse
+> = (
+  this: FastifyInstance<RawServer, RawRequest, RawReply>,
+  req: FastifyRequest<RawRequest>,
+  reply: FastifyReply<RawReply>,
+  done: (err?: Error) => void,
```

Comment:
Thats the best part about TS 😄 make them conform to how the API is supposed to be used!

Reply:
@evanshortiss I've added previously custom error type `FastifyError` where it can't be overridden. And in callbacks it's easy to introduce your own type. 
See https://github.com/fastify/fastify/pull/1250

Reply:
Adding your work now @SkeLLLa